### DISCUSSION TITLE
Brightness

### DISCUSSION THREAD

**Discussion by mindtripper** [Anders-Holst/xled_plus#17] [General]
Is there any way to control the brightness?

This component is elevating x-mas.
Thank you for providing the code and samples!

**Comment by DanDixon** [comment]
I see a function to set the brightness in the original xled implementation but not in xled_plus (so I'm almost certain it's possible)
https://github.com/scrool/xled/blob/develop/xled/control.py#L498

Not the answer you're looking for, but in the meantime, you can set the brightness with the original app, or you can also set the brightness with the Home Assistant integration https://www.home-assistant.io/integrations/twinkly/

Changing the brightness with either of these methods appears to be independent of any effect you've applied with xled_plus

**Reply by mindtripper** [reply]
I'm also a HA user.  The integrations (Twinkly and Adaptive_Lighting) I used to control Twinkly had to be partially disabled. Because they would interfere with some of the sample effects from this repo. Especially the ones that are "continuous effect - press Return to stop it". Any brightness change would halt the "continuous effect - press Return to stop it".
Having these effects is far cooler than what the integrated Twinkly component can do.
This is also why I asked about brightness. As it will allow me to modify the samples, if I decide to, set brightness before they start :)

So, after pip installing this module into my HA instance, I hacked in place a couple of shell_commands and a sensor that start either a random demo effect or one that matches the current day.

```
Shell_Commands:

#Will kill the process that runs "continuous effect - press Return to stop it"
xled_plus_strings_kill: kill -9 $(pgrep -f 'python -m xled_plus.samples')

#Will start the demo effect that matches the current day.
xled_plus_strings_today: nohup bash -c 'python -m xled_plus.samples.{% if now().day <= 30 -%}day{{ now().day }}{%- else -%}{{ ['day31a','day31b'] | random }}{%- endif %} ip.of.twi.kly  > /dev/null 2>&1 &'

#Will start the demo effect that matches the current day.
xled_plus_strings_random_day: nohup bash -c 'python -m xled_plus.samples.{{ ['day1','day2','day3','day4','day5','day6','day7','day8','day9','day10','day11','day12','day13','day14','day15','day16','day17','day18','day19','day20','day21','day22','day23','day24','day25','day26','day27','day28','day29','day30','day31a','day31b','candycane'] | random }} ip.of.twi.kly 2>&1 &'

```

And I have very rudimentary sensor: that will tell me if a "continuous effect - press Return to stop it" is running.

```
- platform: command_line
  name: xled_pid
  command: "pgrep -xf 'python -m xled_plus.samples.*' | grep -o '[[:digit:]]*' || echo '0'"
```

**Reply by Anders-Holst** [reply]
I am happy that you like the continuous effects! Those are what I use mostly myself.

I see. I tried now to write a simple python script to change the brightness. However, it turns out that it too interrupts the continuous effect. It appears that what happens is that when another client connects to the Twinkly device, it disconnects from the ongoing real-time effect. This is regardless of whether the new connection is from the app, HA, or python. Most dissatisfying... So my recommendation is to set the brightness before starting a continuous effect.

By the way: very nice code you have written for selecting the current, or random, daily effect!

**Comment by Anders-Holst** [comment]
Indeed it is possible. As Dan notes, there is a function in xled, and xled_plus inherits from xled (HighControlInterface in xled_plus inherits directly from ControlInterface in xled), so you can call `ctr.set_brightness(value)` with a value between 0 and 100 which is the percent of maximum brightness. This uses the same entry point in the device as the app and Home Assistant (sort of the same but another interface), so it has exactly the same effect, ie all effects are dimmed to the given percentage.

**Reply by mindtripper** [reply]
This saved me hours of trial-and-error scouring/modifying the code.

Thank you!

**Reply by Anders-Holst** [reply]
And in case someone may have use for it, here is my python script for changing the brightness from the command line, without needing the app. (Even though it *will* interrupt an ongoing continuous effect.) Put it in a file `setbrightness.py` and call it as for example `python -m setbrightness 50`

```
"""
Changing brightness from command line.
Run as 'python -m setbrightness PERCENT' where PERCENT is between 0 and 100.
"""

from xled_plus.samples.sample_setup import *

if len(argv) > 1:
    perc = int(argv.pop(1))
else:
    perc = 100
    
ctr = setup_control()

ctr.set_brightness(perc, enabled=True if perc < 100 else False)
```

**Reply by DanDixon** [reply]
set_brightness worked for me as well in a larger script.

Thanks for sharing your shell commands. It gives me some new ideas for what's possible.

I'm using shell commands in Home Assistant to pass the effect name and 1-5 colors to modified scripts. This way the twinkly strings can match the light colors in my house and so it's all controllable from a dashboard.

**Comment by DanDixon** [comment]
I'm not able to make get_brightness work

This works
`ctr.set_brightness(50)`

This doesn't
`bright = ctr.get_brightness()`

I get this value instead of a number
`print (bright)`
`<ApplicationResponse [1000]`

Any thoughts?

**Reply by Anders-Holst** [reply]
It is a function in ControlInterface from the xled package, and by some reason the interface is designed to return an "ApplicationResponse" object, from which you can then reference the specific components, such as e.g `ctr.get_brightness()['value']`. I am considering to add wrappers around all those functions to make them return their values in some readable and more intuitive format like a dict. What I do when calling one of those functions interactively from a python shell is for example `ctr.get_brightness()._data` which gives me the dict with all returned values.
Note though that when called from within other code, the ApplicationResponse object behaves just like a dict, even though you don't see the contents when you print it.

**Reply by DanDixon** [reply]
So as simple as:
```
    brightdata = ctr.get_brightness().data
    bright = brightdata["value"]
```

Thank you. This is very helpful. Makes sense now
